internal/consumer/dlq.go

```diff
@@ -109,4 +148,197 @@ func NewDLQNoop() DLQ {
 	return &dlqNoop{}
 }
 func (d *dlqNoop) Add(m kafka.Message) error { return nil }
-func (d *dlqNoop) Close()                    {}
+func (d *dlqNoop) Stop()                     {}
+func (d *dlqNoop) Start() error              { return nil }
+
+func newBufferedDLQImpl(topic string, producer sarama.SyncProducer, timeLimit <-chan time.Time, sizeLimit <-chan time.Time, tally tally.Scope, log *zap.Logger) *bufferedDLQImpl {
+	return &bufferedDLQImpl{
+		topic:            topic,
+		producer:         producer,
+		timeLimit:        timeLimit,
+		sizeLimit:        sizeLimit,
+		stopC:            make(chan struct{}),
+		closedC:          make(chan struct{}),
+		messageBufferMap: newMessageBufferMap(),
+		errorMap:         newErrorMap(),
+		scope:            tally,
+		logger:           log,
+		lifecycle:        util.NewRunLifecycle(fmt.Sprintf("dlq-%s", topic), log),
+	}
+}
+
+func (d *bufferedDLQImpl) Start() error {
+	return d.lifecycle.Start(func() error {
+		go d.eventLoop()
+		return nil
+	})
+}
+
+func (d *bufferedDLQImpl) Stop() {
+	d.lifecycle.Stop(func() {
+		close(d.stopC)
+		<-d.closedC
+	})
+}
+
+func (d *bufferedDLQImpl) Add(msg kafka.Message) error {
+	key := kafkaMessageKey(msg)
+	sm := d.newSaramaMessage(msg)
+	d.messageBufferMap.Add(key, sm)
+	return d.errorMap.GetAndRemove(key) // errorMap.Get blocks until there is error or nil for this message
+}
+
+func (d *bufferedDLQImpl) eventLoop() {
+	for {
+		select {
+		case <-d.timeLimit:
+			d.flush()
+		case <-d.sizeLimit:
+			d.flush()
+		case <-d.stopC:
+			d.shutdown()
+			return
+		}
+	}
+}
+
+func (d *bufferedDLQImpl) shutdown() {
+	d.flush()
+	close(d.closedC)
+}
+
+func (d *bufferedDLQImpl) flush() {
+	d.messageBufferMap.Lock()
+	defer d.messageBufferMap.Unlock()
+	defer d.messageBufferMap.clear()
+
+	// Get and send buffered messages
+	err := d.producer.SendMessages(d.messageBufferMap.messages())
+
+	// Some messages may fail, so parse failures into errorMap
+	producerErrs, ok := err.(sarama.ProducerErrors)
+	if !ok {
+		d.errorMap.PutAll(d.messageBufferMap.keys(), errors.New("failed to send message to DLQ"))
+		return
+	}
+
+	for _, producerErr := range producerErrs {
+		err := producerErr.Err
+		key, ok := producerErr.Msg.Metadata.(string)
+		if !ok {
+			d.errorMap.Put(key, errors.New("failed to send message to DLQ"))
+			continue
+		}
+		d.errorMap.Put(key, err)
+	}
+
+	// Put nil as default for the errors in errorMap
+	// because cond waits on non-existence of key.
+	d.errorMap.PutAll(d.messageBufferMap.keys(), nil)
+
+	// Broadcast that errorMap has been updated
+	d.errorMap.cond.Broadcast()
```

Comment:
the `Golang` way of doing this is to have a response channel for every caller  - that way you wake up just the goroutine that this response corresponds to rather than all goroutines (broadcast) ie.

```
func Add(msg) error {
   respCh := make(chan error) // must be blocking channel
   sendMsgAsync(msg, respCh) // uses async producer and keeps a map of msgID -> respChannel
   result := <-respCh // block on response
   return result
} 
```